modules/renter/hostdb/hostdb.go

```diff
@@ -35,137 +37,177 @@ const (
+	hdb := &HostDB{
+		cs:         cs,
+		deps:       deps,
+		gateway:    g,
+		persistDir: persistDir,
+
+		scanPool: make(chan modules.HostDBEntry, scanPoolSize),
+	}
 
 	// Create the persist directory if it does not yet exist.
 	err := os.MkdirAll(persistDir, 0700)
 	if err != nil {
 		return nil, err
 	}
+
 	// Create the logger.
 	logger, err := persist.NewFileLogger(filepath.Join(persistDir, "hostdb.log"))
 	if err != nil {
 		return nil, err
 	}
+	hdb.log = logger
 
-	// Create HostDB using production dependencies.
-	return newHostDB(cs, stdDialer{}, stdSleeper{}, newPersist(persistDir), logger)
-}
-
-// newHostDB creates a HostDB using the provided dependencies. It loads the old
-// persistence data, spawns the HostDB's scanning threads, and subscribes it to
-// the consensusSet.
-func newHostDB(cs consensusSet, d dialer, s sleeper, p persister, l *persist.Logger) (*HostDB, error) {
-	// Create the HostDB object.
-	hdb := &HostDB{
-		dialer:  d,
-		sleeper: s,
-		persist: p,
-		log:     l,
-
-		// TODO: should index by pubkey, not ip
-		activeHosts: make(map[modules.NetAddress]*hostEntry),
-		allHosts:    make(map[modules.NetAddress]*hostEntry),
-		scanPool:    make(chan *hostEntry, scanPoolSize),
-	}
-
+	// The host tree is used to manage hosts and query them at random.
 	hdb.hostTree = hosttree.New(hdb.calculateHostWeight)
 
 	// Load the prior persistence structures.
-	err := hdb.load()
+	err = hdb.load()
 	if err != nil && !os.IsNotExist(err) {
 		return nil, err
 	}
 
+	// Don't perform the remaining startup in the presence of a quitAfterLoad
+	// disruption.
+	if hdb.deps.disrupt("quitAfterLoad") {
+		return hdb, nil
+	}
+
 	err = cs.ConsensusSetSubscribe(hdb, hdb.lastChange)
 	if err == modules.ErrInvalidConsensusChangeID {
+		// Subscribe again using the new ID. This will cause a triggered scan
+		// on all of the hosts, but that should be acceptable.
 		hdb.lastChange = modules.ConsensusChangeBeginning
-		// clear the host sets
-		hdb.activeHosts = make(map[modules.NetAddress]*hostEntry)
-		hdb.allHosts = make(map[modules.NetAddress]*hostEntry)
-		// subscribe again using the new ID
 		err = cs.ConsensusSetSubscribe(hdb, hdb.lastChange)
 	}
 	if err != nil {
 		return nil, errors.New("hostdb subscription failed: " + err.Error())
 	}
 
 	// Spin up the host scanning processes.
+	if build.Release != "testing" {
+		go hdb.threadedOnlineCheck()
+	} else {
+		// During testing, the hostdb is just always assumed to be online, since
+		// the online check of having nonlocal peers will always fail.
+		hdb.online = true
+	}
 	for i := 0; i < scanningThreads; i++ {
 		go hdb.threadedProbeHosts()
 	}
-	go hdb.threadedScan()
+
+	// Spawn the scan loop during production, but allow it to be disrupted
+	// during testing. Primary reason is so that we can fill the hostdb with
+	// fake hosts and not have them marked as offline as the scanloop operates.
+	if !hdb.deps.disrupt("disableScanLoop") {
+		go hdb.threadedScan()
+	}
```

Comment:
This is essentially the alternative to the barehostdb. Disable just the features that you need to disable, as opposed to everything.